tachyon/zk/plonk/circuit/assignment.h

```diff
@@ -0,0 +1,119 @@
+  using AnnotationCallback = base::OnceCallback<std::string()>;
+  using AssignCallback = base::OnceCallback<math::RationalField<F>()>;
+
+  // Creates a new region and enters into it.
+  //
+  // Panics if we are currently in a region (if |ExitRegion()| was not called).
+  //
+  // Not intended for downstream consumption; use |Layouter::AssignRegion()|
+  // instead.
+  virtual void EnterRegion(AnnotationCallback annotation) {}
+
+  // Allows the developer to include an annotation for an specific column
+  // within a |Region|.
+  //
+  // This is usually useful for debugging circuit failures.
+  virtual void AnnotateColumn(AnnotationCallback annotation,
+                              const AnyColumn& column) {}
+
+  // Exits the current region.
+  //
+  // Panics if we are not currently in a region (if |EnterRegion()| was not
+  // called).
+  //
+  // Not intended for downstream consumption; use |Layouter::AssignRegion()|
+  // instead.
+  virtual void ExitRegion() {}
+
+  // Enables a selector at the given row.
+  virtual Error EnableSelector(AnnotationCallback annotation,
+                               const Selector& selector, size_t row) {
+    return Error::kNone;
+  }
+
+  // Queries the cell of an instance column at a particular absolute row.
+  //
+  // Returns |Error::kNone| and populates |instance| with the cell's value, if
+  // known.
+  virtual Error QueryInstance(const InstanceColumn& column, size_t row,
+                              Value<F>* instance) {
+    return Error::kNone;
+  }
+
+  // Assign an advice column value (witness).
+  virtual Error AssignAdvice(AnnotationCallback annotation,
+                             const AdviceColumn& column, size_t row,
+                             AssignCallback assign) {
+    return Error::kNone;
+  }
+
+  // Assign a fixed value.
+  virtual Error AssignFixed(AnnotationCallback annotation,
+                            const FixedColumn& column, size_t row,
+                            AssignCallback assign) {
+    return Error::kNone;
+  }
+
+  // Assign two cells to have the same value
+  virtual Error Copy(const AnyColumn& left_column, size_t left_row,
+                     const AnyColumn& right_column, size_t right_row) {
+    return Error::kNone;
+  }
+
+  // Fills a fixed |column| starting from the given |row| with value |assign|.
+  virtual Error FillFromRow(const FixedColumn& column, size_t row,
+                            AssignCallback assign) {
+    return Error::kNone;
+  }
+
+  // Queries the value of the given challenge.
+  //
+  // Returns |Error::kChallengeUnavailable| if the current synthesis phase is
+  // before the challenge can be queried. Otherwise, it populates |value| with
+  // the challenge.
+  virtual Error GetChallenge(const Challenge& challenge, Value<F>* value) {
+    return Error::kChallengeUnavailable;
+  }
+
+  // Creates a new (sub)namespace and enters into it.
+  //
+  // Not intended for downstream consumption; use |Layouter::namespace()|
```

Comment:
Do you know where |Layouter::namespace()| is coming from? I can't find it from both tachyon and halo2 implementation.